fog/ledger/enclave/impl/src/key_image_store.rs

```diff
@@ -0,0 +1,170 @@
+///   ciphertext.len(), ValueSize must be within 255 bytes of ciphertext.len().
+/// - When the lookup misses, we try to obliviously return a buffer of the
+///   normal size. We do this by remembering the ciphertext size byte of the
+///   last stored ciphertext.
+pub struct KeyImageStore<OSC: ORAMStorageCreator<StorageDataSize, StorageMetaSize>> {
+    /// Oblivious map to hold KeyImageStoreRecords
+    omap: Box<<ObliviousMapCreator<OSC> as OMapCreator<KeySize, ValueSize, McRng>>::Output>,
+
+    /// The logger object
+    logger: Logger,
+}
+
+impl<OSC: ORAMStorageCreator<StorageDataSize, StorageMetaSize>> KeyImageStore<OSC> {
+    pub fn new(desired_capacity: u64, logger: Logger) -> Self {
+        Self {
+            omap: Box::new(<ObliviousMapCreator<OSC> as OMapCreator<
+                KeySize,
+                ValueSize,
+                McRng,
+            >>::create(
+                desired_capacity, STASH_SIZE, McRng::default
+            )),
+            logger,
+        }
+    }
+
+    // add a key image containing block index and timestamp
+    pub fn add_record(
+        &mut self,
+        key_image: &KeyImage,
+        block_index: u64,
+        timestamp: u64,
+    ) -> Result<(), AddRecordsError> {
+        let mut value = A8Bytes::<ValueSize>::default();
+        let mut key = A8Bytes::<KeySize>::default(); // key used to add to the oram for key image
+        key.clone_from_slice(&key_image.as_ref());
+        // write block index data to  value[0..8] write the time stamp data to
+        // value[8..16]
+        value[0..8].clone_from_slice(&block_index.to_le_bytes());
+        value[8..16].clone_from_slice(&timestamp.to_le_bytes());
+        // Note: Passing true means we allow overwrite, which seems fine since
+        // the vasearch_key: &[u8]lue is not changing
+        let omap_result_code = self.omap.vartime_write(&key, &value, Choice::from(1));
+        if omap_result_code == OMAP_INVALID_KEY {
+            return Err(AddRecordsError::KeyWrongSize);
+        } else if omap_result_code == OMAP_OVERFLOW {
+            return Err(AddRecordsError::MapOverflow(
+                self.omap.len(),
+                self.omap.capacity(),
+            ));
+        } else if omap_result_code == OMAP_FOUND {
+            log::debug!(self.logger, "An omap key was found");
+        } else if omap_result_code != OMAP_NOT_FOUND {
+            panic!(
+                "omap_result_code had an unexpected value: {}",
+                omap_result_code
+            );
+        }
+        Ok(())
+    }
+
+    // return new struct KeyImageResult which contains block index and timestamp of
+    // key image key image as ref to convert key image to 32 bits,
+    // call the oram to query to to key image data
+    pub fn find_record(&mut self, key_image: &KeyImage) -> KeyImageResult {
+        let mut result = KeyImageResult {
+            key_image: KeyImage::from(2),
+            spent_at: 0u64,
```

Comment:
i think this should default to `u64::MAX`

Reply:
the basic reason is, that matches the previous behavior.

the reason for the previous behavior is, if a protobuf field is 0, that will make the encoded protobuf shorter, and that is an information leak